compiler/rustc_middle/src/ty/util.rs

```diff
@@ -1267,3 +1267,83 @@ pub fn provide(providers: &mut ty::query::Providers) {
     *providers =
         ty::query::Providers { normalize_opaque_types, is_doc_hidden, is_intrinsic, ..*providers }
 }
+
+/// Returns `true` if nullable pointer optimizations apply to the given type,
+/// e.g. `Option<NonNull<_>>` as well as `Option<T>`s where `T` is a transparent
+/// newtype around another niche-able type.
+///
+/// Currently supports function pointers (`fn(...) -> ..`), `Box<_>`, references, `NonNull<_>`,
+/// `NonZero*` and `#[repr(transparent)]` newtypes around them
+pub fn nonnull_opt_applies<'tcx>(tcx: TyCtxt<'tcx>, ty: Ty<'tcx>, is_definition: bool) -> bool {
```

Comment:
This is somehow slightly incorrect, it looks like previously a `true` result was being treated as a "maybe" instead of a canonical "yes, this is can be null pointer opt'ed"
Specifically, it fails in `std/src/io/error/repr_bitpacked.rs` on a size assertion that `Option<Repr>` is 8 bytes where `Repr` is
```rust
#[repr(transparent)]
struct Repr(NonNull<()>, PhantomData<ErrorData<Box<Custom>>>);
```
Which confuses me since it *should* work?
- `nonnull_opt_applies()` gets an `Option<Repr>` and sees that `Option` is `Option`-shaped
- It gets the inner type, `Repr`
- It passes `Repr` to `ty_is_known_nonnull()`
  - `ty_is_known_nonnull()` sees that `Repr` is a `#[transparent]` ADT which is not a union
  - `Repr` doesn't have `#[rustc_nonnull_optimization_guaranteed]` but that's ok, we continue
  - `Repr` doesn't have `repr(no_niche)` but that's ok, we continue
  - We find the non-zst field of `Repr` which is `NonNull<()>`
    - We call `ty_is_known_nonnull()` on `NonNull<()>`
    - `NonNull<()>` is a transparent ADT which isn't a union
    - `NonNull<()>` has `#[nonnull_optimization_guaranteed]`, we return true
  - `.any(|field| ty_is_known_nonnull(field))` gets a `true` from `NonNull<()>` and returns `true`
- `ty_is_known_nonnull(Repr)` returns `true`